### PR TITLE
feat: add support for rich language using ipset in destination

### PR DESCRIPTION
**Pull request by derekdai**
See #706

### CODE REVIEW

**Comment by erig0** [comment]
Hi, Thanks for working on this. It looks like there are some test failures. Can you fix them?

**Comment by derekdai** [comment]
I am working on it.

**Comment by derekdai** [comment]
Hi @erig0, I think this PR can be reviewed now.

File: src/firewall/core/ipXtables.py [hunk 1]
```diff
@@ -1116,15 +1116,22 @@ def _rich_rule_destination_fragment(self, rich_dest):
...
+            if rich_dest.invert:
+                rule_fragment.append("!")
+            if check_single_address("ipv6", rich_dest.addr):
+                rule_fragment += [ "-d", normalizeIP6(rich_dest.addr) ]
+            elif check_address("ipv6", rich_dest.addr):
+                addr_split = rich_dest.addr.split("/")
+                rule_fragment += [ "-d", normalizeIP6(addr_split[0]) + "/" + addr_split[1] ]
+            else:
+                rule_fragment += [ "-d", rich_dest.addr ]
+        elif hasattr(rich_dest, "ipset") and rich_dest.ipset:
```
Comment (erig0) [inline src/firewall/core/ipXtables.py:1129]: nit: `hasattr()` in unnecessary as you guarantee `ipset` exists when creating the `Rich_Destination` instance.

File: src/firewall/core/nftables.py [hunk 2]
```diff
@@ -1245,7 +1245,12 @@ def _rich_rule_family_fragment(self, rich_family):
     def _rich_rule_destination_fragment(self, rich_dest):
         if not rich_dest:
             return {}
-        return self._rule_addr_fragment("daddr", rich_dest.addr, invert=rich_dest.invert)
+        if rich_dest.addr:
+            address = rich_dest.addr
+        elif hasattr(rich_dest, "ipset") and rich_dest.ipset:
```
Comment (erig0) [inline src/firewall/core/nftables.py:1250]: same comment as above.. `hasattr()` is unnecessary.

**Review by erig0 — COMMENTED** [review]
This looks great. The only thing missing is test coverage. If you don't want to add then I can manually pull your PR and add that myself.

File: src/firewall/core/io/zone.py [hunk 3]
```diff
@@ -199,6 +199,11 @@ def _check_config(self, config, item, all_config):
                 if not checkIPnMask(source) and not checkIP6nMask(source) and \
                    not check_mac(source) and not source.startswith("ipset:"):
                     raise FirewallError(errors.INVALID_ADDR, source)
+        elif item == "destination":
+            for destination in config:
+                if not checkIPnMask(destination) and not checkIP6nMask(destination) and \
+                   not destination.startswith("ipset:"):
+                    raise FirewallError(errors.INVALID_ADDR, destination)
```
Comment (erig0) [inline src/firewall/core/io/zone.py:206]: This should actually be `INVALID_DESTINATION`.

Comment (erig0) [reply src/firewall/core/io/zone.py:206]: Actually this whole block is not needed. This part of `check_config` applies to zone based elements. Not rich rules.

**Comment by erig0** [comment]
Thanks! This is now on master.

**Comment by derekdai** [comment]
Thank you for adding tests and all the changes, so I can learn from it too.

**Pull request closed without merging**